### PR TITLE
use a simpler calculation in the SolarGaugeCard

### PR DESCRIPTION
**Pull request by eringerli**
## Proposed change
In the tooltip, the formula is described simply as ratio between used and fed in solar energy, but the calculations included the battery storage. That formula reduced the self consumption ratio artificially if the batteries are not 100% efficient, which they can't ever be.

This commit uses the following formula:

`100% - (total grid export / total solar production) * 100%`

This is more logical (you can calculate this ratio easily for yourself with only the displayed data) and is now consistent with the tooltip. This tripped me, when I tried to calculate the ratio for myself and got different values than displayed in the dashboard. My storage system is about 85% efficient (as they mostly are), so the gauge was consistently about 10% too low. Also, I think this gauge should be specific to one day and shouldn't include stored energy, as the tooltip is right: this should be an indicator if controllable loads and a storage system are a good idea or not.

I also adjusted the tests and prohibited for division-by-zero errors.

## Breaking change
This PR propopses a different formula for the solar self consumption gauge. I'm not sure if this constitutes as "Breaking Change". It generally goes up compared to the original formula if you have a storage system, for all other cases it stays the same.

## Screenshots
Bildschirmfoto_20260404_124839

## Type of change
- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Checklist
- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

### CODE REVIEW

**Comment by karwosts** [comment]
Can you review the original PR that added this logic, and the associated issues it closed, and make sure you aren't just bringing back old issues?

I feel like we maybe used to have this formula and changed it to try to fix some issue, though I don't recall exactly what it was.

But we probably shouldn't just delete large swathe of code without understanding what it was for.

**Comment by eringerli** [comment]
Thanks for the suggestions, I will look into it. I will add a test for this and adjust the formula accordingly. I set the PR to draft in the meantime.

If I can make a suggestion: next time, please add the description of the problem as comment into the code, instead of what the code does, then I would have tested it already.

**Comment by eringerli** [comment]
That PR was #25326

**Comment by eringerli** [comment]
@karwosts I think, this works now. I still have to adjust the test suite and experiment some more, but I think the solution with the aggregated data from `computeConsumptionData()` is good.

**Comment by eringerli** [comment]
@karwosts I adjusted the tests. I'm not that experienced with test and added `// @ts-expect-error TS2740` to the `consumption` parameter. If this is not wanted, please describe how to change that

File: src/data/energy.ts [hunk 1]
```diff
@@ -1357,88 +1357,15 @@ export const formatConsumptionShort = (
...
-      });
-    }
-    if (consumption.solar_to_battery[t]) {
-      batteryLifo.push({
-        type: "solar",
-        value: consumption.solar_to_battery[t],
-      });
-    }
-
-    let batteryToGrid = consumption.battery_to_grid[t] ?? 0;
```
Comment (MindFreeze) [inline src/data/energy.ts:1401]: This is no longer tracked

Comment (eringerli) [reply src/data/energy.ts:1401]: It's tracked in `sum.total.solar` and `consumption.total.solar_to_grid`, so the ratio is the effective solar grid export (explicitly without the export from storage) to the total solar production. 

For me, the solar gauge should show exactly what the tooltip says: how much solar energy was used (the `solar_to_grid` is the opposite of that, hence `1 - ( ... ) * 100`) relative to the total production over a certain period if time. If you factor in any storage, then it becomes dependant of the efficiency of that, which gives too low values. Also, storage can be filled with energy from the grid (dependant on the rules about storage systems in your country and if you have dynamic pricing), this shouldn't be shown as 'solar'. If you want to account for that, the algorythm gets even more complicated, and still wouldn't be as described in the tooltip.

Comment (MindFreeze) [reply src/data/energy.ts:1401]: `battery_to_grid` is not included in `solar_to_grid` so it is just not tracked and will be treated as self-consumed when it is not. This looks like a regression to me.
This does NOT account for the following scenario (which the old LIFO algorithm explicitly handled): solar charges the battery, then the battery later exports to the grid. In that case, the solar energy was not really "self-consumed" -- it went solar -> battery -> grid. The old algorithm tracked this. The new formula misses it, counting that energy as self-consumed because it was not directly exported from solar to grid.
The previous code was a estimation and not 100% correct but this looks even less correct.

Comment (MindFreeze) [reply src/data/energy.ts:1401]: I understand the need for simplicity if you don't have `battery_to_grid` which is the most common case, but if you do have `battery_to_grid`, we shouldn't just ignore it.

**Comment by eringerli** [comment]
The dark one is the original algorythm, the light one my proposal. On this day, there was about 6% grid export of solar, but the gauge shows only 84%. This is clearly wrong.

This difference is apparent, if the stored energy is much higher than the retrieved energy. If they are about the same, the difference gets smaller.

Bildschirmfoto_20260407_092329
Bildschirmfoto_20260407_090251
Bildschirmfoto_20260407_092015

File: src/data/energy.ts (hunk 1 above)
Comment (eringerli) [reply src/data/energy.ts:1401]: In my opinion, the solar gauge is explicitly not about storage, but about the ratio between used and total solar energy. It shouldn't change if that energy is later fed into the grid.

I did a small research on the web, and all other energy systems don't factor in any stored energy into this value, even in the research papers I found. They all do the most obvious thing: _(solar production - solar grid export) / (solar production)_. I think, we shoudn't show different values than all the other solutions on the market. The Self Sufficiency Gauge and the Net Import/Export Gauge should show if you exported more than you used, not the Solar Usage Gauge.

**Comment by MindFreeze** [comment]
TBH I would just remove this card if it was up to me, or change it to count only how much solar you consumed directly without it going to batteries or the grid. Anything else is too complicated and unclear.
Changing the behavior like this PR proposes doesn't look like an improvement to me. It will probably spawn a bunch of new issues and discussions that are not worth it.

**Comment by eringerli** [comment]
> TBH I would just remove this card if it was up to me, or change it to count only how much solar you consumed directly without it going to batteries or the grid. Anything else is too complicated and unclear. 

No, this indicator is important. But the original algorythm is way too complicated and undocumented.

> Changing the behavior like this PR proposes doesn't look like an improvement to me. It will probably spawn a bunch of new issues and discussions that are not worth it.

This PR does exactly as in the tooltip and the documentation, so I don't really see, how this opens new issues? The original behaviour was incorrect, so it's not a regression, it's a bugfix.

Here is the documentation from https://www.home-assistant.io/dashboards/energy/#solar-consumed-gauge:
Bildschirmfoto_20260407_174225

**Comment by eringerli** [comment]
The original formula replaced this: https://github.com/home-assistant/frontend/pull/25326/changes#diff-f9ea1f6fc6479cd644051e5bacffcfffe62466af146c4a0f07c6535e0266bbebL86

My PR uses only the energy directly fed in to the grid from solar, so this solves the problem too and more cleanly IMHO, as `computeConsumptionData()` is pretty precise and is used in other places.

**Comment by MindFreeze** [comment]
I don't understand why you think it's not consistent with the tooltip or docs. They say nothing about storage specifically. They talk about usage. So the question is if you charge a battery and later export that energy to the grid, was it used by the home? You seem to think yes, but I don't agree. Why does it matter that the energy passed through the battery on its way out? It was just buffered. It left the home so it was not used by the home.
And why does the efficiency of the battery matter here? Any energy lost in the battery will not be counted as exported anyway.

**Comment by NoRi2909** [comment]
It looks to me like the only issue here is the name of that gauge:
"Self-_consumed_ solar energy"

As we have discussed in other PRs "consumed" should only be applied to energy used by the loads in the house. As long as it is stored in a battery that isn't really "consumption", yet.

So if we simply change that label to "Self-_used_ solar energy" things become much clearer.

In German this would then change the translation from "PV-Eigenverbrauch" to "PV-Eigennutzung" or "Selbstgenutzte PV-Energie".

NOTE: Over the course of a year there is really not much of a difference between "self-consumed" or "self-used" as the battery's influence becomes negligible. But if you look at a single day there can be a big difference between the two.

**Comment by eringerli** [comment]
@NoRi2909 Thanks for your thoughts on the matter.

Do you support my PR or not?

I think, this gauge should represent the same value as other solutions on the market and in the literature: it is a measure of how much of your solar energy leaves your home relative to the total solar production, without consideration when or how you use it. I would suggest to use the much simpler formula in this PR instead of the more complex original algorythm. Especially as this PR uses already existing and tested functions (all other energy dashboard gauges use them too) to aggregate the data into totals instead of code used only in one location.

**Comment by NoRi2909** [comment]
@eringerli I have no opinion on your PR, I just wanted to propose a solution that might help in getting both sides together by just slightly re-labeling this gauge.

As I did the German translation I understand that "PV-Eigenverbrauch" is probably even more misleading while being a correct translation of the current English "Self-consumed solar energy".

**Comment by laupalombi** [comment]
Hey hello! Seems like no UX is needed, the PR discussion is about the algorythm, and the representation looks fine to me. I'll remove the Needs UX label.

**Comment by MindFreeze** [comment]
Alright I'm closing this. Seems to be very subjective and unclear so I would prefer to not delete a bunch of code that would not be trivial to reproduce.
As stated before, I'm in favor of removing this card from the energy dashboard.
What would make more sense to show is used energy (excluding battery charging) relative to produced from solar. But that overlaps with the Self-sufficiency card.
This PR would be great when there is no battery, but when there is, we can't just pretend it doesn't exist.
